PyPDF2/_writer.py

```diff
@@ -196,16 +202,23 @@ def pdf_header(self) -> bytes:
     def pdf_header(self, new_header: bytes) -> None:
         self._header = new_header
 
-    def _add_object(self, obj: Optional[PdfObject]) -> IndirectObject:
+    def _add_object(self, obj: PdfObject) -> IndirectObject:
+        if hasattr(obj, "indirect_ref") and obj.indirect_ref.pdf == self:  # type: ignore
+            return obj.indirect_ref
         self._objects.append(obj)
-        return IndirectObject(len(self._objects), 0, self)
+        obj.indirect_ref = IndirectObject(len(self._objects), 0, self)
+        return obj.indirect_ref
 
-    def get_object(self, ido: IndirectObject) -> PdfObject:
+    def get_object(self, ido: Union[int, IndirectObject]) -> PdfObject:
+        if isinstance(ido, int):
+            return self._objects[ido - 1]
         if ido.pdf != self:
             raise ValueError("pdf must be self")
         return self._objects[ido.idnum - 1]  # type: ignore
 
-    def getObject(self, ido: IndirectObject) -> PdfObject:  # pragma: no cover
+    def getObject(
+        self, ido: Union[int, IndirectObject]
+    ) -> PdfObject:  # pragma: no cover
```

Comment:
This is a change that is not related to cloning, right? It seems to be an unrelated improvement you made while writing this PR.

As this PR is rather big and might take a while to review, I would like to move this to a separate PR ("Miscellaneous fixes" or similar)